File: frontend/src/api/usuario.service.ts
// Capa de API de Usuarios (incluye roles, bloqueo y contrasena).
import http from '@/api/http'
import type {
  CambioPasswordDto,
  UsuarioDetalleDto,
  UsuarioDto,
  UsuarioRolDto,
} from '@/types/usuario.type'

export function listarUsuarios() {
  return http.get<UsuarioDetalleDto[]>('/usuarios/listar').then((r) => r.data)
}

export function crearUsuario(dto: UsuarioDto) {
  return http.post<UsuarioDetalleDto>('/usuarios/crear', dto).then((r) => r.data)
}

export function actualizarUsuario(idUsuario: number, dto: UsuarioDto) {
  return http
    .put<UsuarioDetalleDto>('/usuarios/actualizar', dto, { params: { idUsuario } })
    .then((r) => r.data)
}

export function eliminarUsuario(idUsuario: number) {
  return http.delete('/usuarios/eliminar', { params: { idUsuario } })
}

export function cambiarBloqueo(idUsuario: number, bloqueado: boolean) {
  return http
    .patch<UsuarioDetalleDto>('/usuarios/bloqueo', null, { params: { idUsuario, bloqueado } })
    .then((r) => r.data)
}

export function cambiarPassword(idUsuario: number, dto: CambioPasswordDto) {
  return http.patch('/usuarios/password', dto, { params: { idUsuario } })
}

export function asignarRol(dto: UsuarioRolDto) {
  return http.post<UsuarioDetalleDto>('/usuarios/asignar-rol', dto).then((r) => r.data)
}

export function quitarRol(dto: UsuarioRolDto) {
  return http.post<UsuarioDetalleDto>('/usuarios/quitar-rol', dto).then((r) => r.data)
}

File: src/main/java/com/uap/control_tickets/services/interfaces/UsuarioService.java
package com.uap.control_tickets.services.interfaces;

import com.uap.control_tickets.dto.usuario.CambioPasswordDto;
import com.uap.control_tickets.dto.usuario.UsuarioDetalleDto;
import com.uap.control_tickets.dto.usuario.UsuarioDto;
import com.uap.control_tickets.dto.usuario.UsuarioRolDto;

import java.util.List;

/** Contrato del CRUD de Usuario y la gestion de sus roles. */
public interface UsuarioService {
    List<UsuarioDetalleDto> listar();
    UsuarioDetalleDto obtener(Long idUsuario);
    UsuarioDetalleDto miPerfil();
    UsuarioDetalleDto crear(UsuarioDto dto);
    UsuarioDetalleDto actualizar(Long idUsuario, UsuarioDto dto);
    void eliminar(Long idUsuario);

    // Gestion de estado y credenciales
    UsuarioDetalleDto cambiarBloqueo(Long idUsuario, boolean bloqueado);
    void cambiarPassword(Long idUsuario, CambioPasswordDto dto);

    // Gestion de roles del usuario
    UsuarioDetalleDto asignarRol(UsuarioRolDto dto);
    UsuarioDetalleDto quitarRol(UsuarioRolDto dto);
}
